server/join_test.go

```diff
@@ -85,33 +78,252 @@ func waitMembers(svr *Server, c int) error {
 	return errors.New("waitMembers Timeout")
 }
 
-func (s *testJoinServerSuite) TestJoin(c *C) {
-	svrs := make([]*Server, 0, len(s.cfgs))
-	for i, cfg := range s.cfgs {
-		svr, err := NewServer(cfg)
+func waitLeader(svrs []*Server) error {
+	// maxRetryTime * waitInterval = 10s
+	maxRetryCount := 20
+	waitInterval := 500 * time.Millisecond
+	for count := 0; count < maxRetryCount; count++ {
+		for _, s := range svrs {
+			// TODO: a better way of finding leader.
+			if s.etcd.Server.Leader() == s.etcd.Server.ID() {
+				return nil
+			}
+		}
+		time.Sleep(waitInterval)
+	}
+	return errTimeout
+}
+
+// notice: cfg has changed
+func startPdWith(cfg *Config) (*Server, error) {
+	svrCh := make(chan *Server)
+	errCh := make(chan error)
+	go func() {
+		svr, err := CreateServer(cfg)
+		if err != nil {
+			errCh <- errors.Trace(err)
+			return
+		}
+		err = svr.StartEtcd(nil)
+		if err != nil {
+			errCh <- errors.Trace(err)
+			return
+		}
+		svrCh <- svr
+		svr.Run()
+	}()
+
+	timer := time.NewTimer(5 * time.Second)
+	defer timer.Stop()
+
+	select {
+	case s := <-svrCh:
+		return s, nil
+	case e := <-errCh:
+		return nil, errors.Trace(e)
+	case <-timer.C:
+		return nil, errTimeout
+	}
+}
+
+type cleanUpFunc func()
+
+func mustNewJoinCluster(c *C, num int) ([]*Config, []*Server, cleanUpFunc) {
+	dirs := make([]string, 0, num)
+	svrs := make([]*Server, 0, num)
+	cfgs := newTestMultiJoinConfig(num)
+
+	for _, cfg := range cfgs {
+		dirs = append(dirs, cfg.DataDir)
+		svr, err := startPdWith(cfg)
 		c.Assert(err, IsNil)
-		defer svr.Close()
 		svrs = append(svrs, svr)
+	}
 
-		go svr.Run()
+	waitMembers(svrs[rand.Intn(num)], num)
 
-		// Make sure new pd is started.
-		err = waitMembers(svrs[0], i+1)
-		c.Assert(err, IsNil)
+	// clean up
+	clean := func() {
+		for _, s := range svrs {
+			s.Close()
+		}
+	}
+
+	return cfgs, svrs, clean
+}
+
+func alive(target, peer *Server) error {
+	timer := time.NewTimer(10 * time.Second)
+	defer timer.Stop()
+
+	ch := make(chan error)
+	go func() {
+		// put something to cluster
+		key := fmt.Sprintf("%d", rand.Int63())
+		value := key
+		client := peer.GetClient()
+		ctx, cancel := context.WithTimeout(context.Background(), requestTimeout)
+		defer cancel()
+		_, err := client.Put(ctx, key, value)
+		if err != nil {
+			ch <- errors.Trace(err)
+			return
+		}
+
+		client = target.GetClient()
+		ctx, cancel = context.WithTimeout(context.Background(), requestTimeout)
+		defer cancel()
+		resp, err := client.Get(ctx, key)
+		if err != nil {
+			ch <- errors.Trace(err)
+			return
+		}
+		if string(resp.Kvs[0].Value) != value {
+			msg := fmt.Sprintf("not match, got: %s, expect: %s", resp.Kvs[0].Value, value)
+			err = errors.New(msg)
+			ch <- err
+			return
+		}
+		ch <- nil
+	}()
+
+	select {
+	case err := <-ch:
+		return err
+	case <-timer.C:
+		return errTimeout
 	}
+}
+
+func (s *testJoinServerSuite) TestJoinCase1(c *C) {
```

Comment:
what's case 1 here? user must re-see the explanation again in another file.
Use detailed name here.